### PR TITLE
[ADD] web_widget_text_count

### PR DESCRIPTION
**Pull request by i-vyshnevska**
Field TextCount, combined to widget text_count, is displayed as a usual text field with a character counter next to it.

Screenshot

Revoked on https://github.com/OCA/web/pull/1214

### CODE REVIEW

**Comment by pedrobaeza** [comment]
The module should be now called `web_widget_text_count`

File: web_widget_text_count/readme/CONTRIBUTORS.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+* Yannick Vaucher <yannick.vaucher@camptocamp.com>
```
Comment (simahawk) [inline web_widget_text_count/readme/CONTRIBUTORS.rst:1]: @i-vyshnevska what about us? :wink:

File: web_widget_text_count/readme/ROADMAP.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+Patch Text class on Python level to allow propagation of size same as in Char fields.
+After modify js to relies it on field attribute.
```
Comment (simahawk) [inline web_widget_text_count/readme/ROADMAP.rst]: ~After modify js to relies it on field attribute.~
"At the moment the `size` attribute can be applied only as an attribute on the `<field`/>` tag.

File: web_widget_text_count/readme/ROADMAP.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+
+Impossible to put counter field after Text field in one line in UI.
```
Comment (simahawk) [inline web_widget_text_count/readme/ROADMAP.rst]: ~Impossible to put counter field after Text field in one line in UI.~
Support inline position of the counter next to the textarea.

File: web_widget_text_count/static/src/js/fields_text_count.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+odoo.define('text_count', function(require) {
+    "use strict";
+
+    require('web.dom_ready');
```
Comment (simahawk) [inline web_widget_text_count/static/src/js/fields_text_count.js]: I don't think dom.ready is needed here.

File: web_widget_text_count/static/src/js/fields_text_count.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        }),
+        supportedFieldTypes: ['text'],
+
+        start: function () {
+            
+            var self = this;
+            return this._super().then(function(){
+                if (self.mode === 'edit') {
+                    if (self.attrs.size) {
+                        // this place can be improved should relay on self.field.size
```
Comment (simahawk) [inline web_widget_text_count/static/src/js/fields_text_count.js]: this comment can be removed. In any case is good to support size on demand per view.

File: web_widget_text_count/static/src/scss/field_text_count.scss [hunk 6]
```diff
@@ -0,0 +1,13 @@
+.text-counter {
```
Comment (simahawk) [inline web_widget_text_count/static/src/scss/field_text_count.scss]: wrap the rules inside the main class of the widget

**Review by simahawk — CHANGES_REQUESTED** [review]

File: web_widget_text_count/readme/ROADMAP.rst (hunk 2 above)
Comment (asaunier) [reply web_widget_text_count/readme/ROADMAP.rst]: @simahawk You can use the ``+-`` button ("Insert a suggestion") in the comments toolbar

File: web_widget_text_count/static/src/scss/field_text_count.scss [hunk 7]
```diff
@@ -0,0 +1,12 @@
+.text-counter-widget{
+        background-color: transparent;
+        float: right;
+        white-space: pre-wrap;
+        border: 1px solid #ccc;
+        border-radius: 3px;
+        padding: 2px 4px;
+        color: #1f1f1f;
+    div.text-counter-widget {
```
Comment (simahawk) [inline web_widget_text_count/static/src/scss/field_text_count.scss:9]: hmm text-counter-widget inside text-counter-widget? 
You can add your own class if needed (there's a param on widget class for it)

File: web_widget_text_count/static/src/js/fields_text_count.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+    var registry = require('web.field_registry');
+    var basicFields = require('web.basic_fields');
+    var FieldText = basicFields.FieldText;
+
+    var FieldTextCount = FieldText.extend({
+        events: _.extend({}, FieldText.prototype.events, {
+            'input': 'count_char',
+        }),
+        supportedFieldTypes: ['text'],
```
Comment (daramousk) [inline web_widget_text_count/static/src/js/fields_text_count.js:12]: Is this necessary?

File: web_widget_text_count/static/src/js/fields_text_count.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+                var size = 0
+                if (self.mode === 'edit') {
+                    if (self.attrs.size) {
+                        self.$el.attr('maxlength', parseInt(self.attrs.size));
+                        size = self.attrs.size - self.$el.val().length;
+                    }
+                    else{
+                        size = self.$el.val().length;
+                    }
+                    self.$el = self.$el.add($('<input class="text-counter-widget" readonly="readonly" value="'+size+'"/>'));
```
Comment (daramousk) [inline web_widget_text_count/static/src/js/fields_text_count.js:26]: Why an `input` and not a `span`  with `o_readonly_modifier` class? The counter looks like it is accepting input when in fact it does not.

**Review by daramousk — CHANGES_REQUESTED** [review]

**Comment by leemannd** [comment]
@simahawk @daramousk do you have time to update your reviews on this topic?
Honestly we are out of time to address them. The widget is currently working well since `2019`. My current proposition would be to merge this in order to let also other people use it.

**Comment by bosd** [comment]
@i-vyshnevska What is the status of this pr?
Can you resolve the changes? Or are they resolved?

(Would like forward port this one to V14 if this one has been approved)

**Comment by bosd** [comment]
@simahawk Can you review/update changes?

**Comment by yvaucher** [comment]
@bosd did you forward port this?
As @leemannd stated we can merge this in the current state and give improvement later.

The main issue I see with this model is the `<input>` that force the user to tab into it on output and should be fixed, otherwise it works.

**Comment by yvaucher** [comment]
Reopennning to merge, fixup will follow.

**Comment by bosd** [comment]
> did you forward port this?

No, not yet. Thanks for coming back to this one.

**Comment by leemannd** [comment]
@yvaucher If you:
- rebase the PR on 12.0
- launch locally the pre-commit
It will be fixed.

I have a PR to show it: https://github.com/OCA/web/pull/2858

**Pull request closed without merging**